File: r1-spike/ts/src/kernel/numeric.ts
// Numeric profile v1: xoshiro128** 1.1 (Blackman/Vigna, public domain), uniform, divide.
import { Fault, SAFE_INT, isInt } from './codec.ts';
import type { Json } from './codec.ts';

const U32 = 0xffffffff;
const TWO32 = 0x100000000;

function checkWords(words: Json): number[] {
  if (!Array.isArray(words) || words.length !== 4) throw new Fault('invalid_rng_state');
  for (const v of words) if (!isInt(v) || v < 0 || v > U32) throw new Fault('invalid_rng_state');
  const w = words as number[];
  if (w.every((v) => v === 0)) throw new Fault('invalid_rng_state');
  return w;
}

const rotl = (v: number, k: number): number => ((v << k) | (v >>> (32 - k))) >>> 0;

function next(w: number[]): [number, number[]] {
  let [a, b, c, d] = w;
  const result = Math.imul(rotl(Math.imul(b, 5) >>> 0, 7), 9) >>> 0;
  const t = (b << 9) >>> 0;
  c = (c ^ a) >>> 0;
  d = (d ^ b) >>> 0;
  b = (b ^ c) >>> 0;
  a = (a ^ d) >>> 0;
  c = (c ^ t) >>> 0;
  d = rotl(d, 11);
  return [result, [a, b, c, d]];
}

export function rngNext(words: Json): [number, number[]] {
  return next(checkWords(words));
}

export function uniform(words: Json, bound: Json, maxDraws: Json): [number, number[]] {
  if (!isInt(bound) || bound < 1 || bound > TWO32) throw new Fault('invalid_bound');
  if (!isInt(maxDraws) || maxDraws < 0) throw new Fault('invalid_rng_budget');
  let w = checkWords(words);
  const limit = TWO32 - (TWO32 % bound);
  for (let i = 0; i < maxDraws; i++) {
    const [raw, nw] = next(w);
    w = nw;
    if (raw < limit) return [raw % bound, w];
  }
  throw new Fault('rng_budget_exhausted');
}

/** Truncating division; exact because % on doubles is exact and (a - r) / b divides evenly. */
export function divide(a: Json, b: Json): [number, number] {
  if (!isInt(a) || !isInt(b) || Math.abs(a) > SAFE_INT || Math.abs(b) > SAFE_INT) throw new Fault('integer_out_of_range');
  if (b === 0) throw new Fault('divide_by_zero');
  const r = a % b;
  const q = (a - r) / b;
  return [q + 0, r + 0];
}

File: r1-spike/elixir/lib/loka_r1/numeric.ex
defmodule LokaR1.Numeric do
  @moduledoc """
  xoshiro128** 1.1 (Blackman/Vigna, public domain; adapted per
  `conformance/numeric-profile.md`), bounded uniform draws and signed
  truncating division. Errors are `{:error, code}` with the model's codes.
  """
  import Bitwise

  @u32 0xFFFFFFFF
  @safe 9_007_199_254_740_991

  @spec rng_next(term()) :: {:ok, non_neg_integer(), [non_neg_integer()]} | {:error, binary()}
  def rng_next(words) do
    if valid_state?(words), do: next(words), else: {:error, "invalid_rng_state"}
  end

  defp next([a, b, c, d]) do
    result = rotl(b * 5 &&& @u32, 7) * 9 &&& @u32
    t = b <<< 9 &&& @u32
    c = bxor(c, a)
    d = bxor(d, b)
    b = bxor(b, c)
    a = bxor(a, d)
    c = bxor(c, t)
    d = rotl(d, 11)
    {:ok, result, [a, b, c, d]}
  end

  defp rotl(v, n), do: (v <<< n ||| v >>> (32 - n)) &&& @u32

  defp valid_state?(words) do
    is_list(words) and length(words) == 4 and
      Enum.all?(words, &(is_integer(&1) and &1 in 0..@u32)) and Enum.any?(words, &(&1 != 0))
  end

  @doc "Uniform integer in [0, bound); errors checked in the model's order."
  @spec uniform(term(), term(), term()) :: {:ok, non_neg_integer(), list()} | {:error, binary()}
  def uniform(words, bound, max_draws) do
    cond do
      not (is_integer(bound) and bound in 1..(1 <<< 32)) -> {:error, "invalid_bound"}
      not (is_integer(max_draws) and max_draws >= 0) -> {:error, "invalid_rng_budget"}
      not valid_state?(words) -> {:error, "invalid_rng_state"}
      true -> draw(words, bound, (1 <<< 32) - rem(1 <<< 32, bound), max_draws)
    end
  end

  defp draw(_words, _bound, _limit, 0), do: {:error, "rng_budget_exhausted"}

  defp draw(words, bound, limit, left) do
    {:ok, raw, words} = next(words)
    if raw < limit, do: {:ok, rem(raw, bound), words}, else: draw(words, bound, limit, left - 1)
  end

  @spec divide(term(), term()) :: {:ok, integer(), integer()} | {:error, binary()}
  def divide(a, b) do
    cond do
      not (is_integer(a) and is_integer(b) and abs(a) <= @safe and abs(b) <= @safe) ->
        {:error, "integer_out_of_range"}

      b == 0 ->
        {:error, "divide_by_zero"}

      true ->
        # Elixir's div/rem already truncate toward zero.
        {:ok, div(a, b), rem(a, b)}
    end
  end
end
